### PR TITLE
Allow redirection for non-safe methods

### LINKED ISSUE #24681: HTTP 307 temporary redirection fails when invoking POST method
**Description:**
On Ballerina HTTP client when a POST request is invoked if the server responds with a HTTP 307 status code, the redirection does not happen from the HTTP client. 

**Steps to reproduce:**
This can be reproduced using the [BBE on redirects](https://ballerina.io/learn/by-example/http-redirects.html).

**Affected Versions:**
Ballerina Swan Lake Preview 1

**OS, DB, other environment details and versions:**
N/A

**Related Issues (optional):**
N/A

**Suggested Labels (optional):**
N/A

**Suggested Assignees (optional):**
N/A

### PR DESCRIPTION
**Pull request by chamil321**
## Purpose

> Existing implementation serves auto redirection only for GET and HEAD methods as there was a requirement of getting user consent before performing redirection for non-safe HTTP methods in rfc2616. But that requirement has been dropped in rfc7231 and if the user enables redirect config in the client configuration, then it also implies the consent as the redirection is disabled in HTTP client by default.

> Also, rfc7231 and rfc7538 specify that 307 and 308 should not allow changing the request method from POST to GET. That means a POST request should be repeated using another POST request during the redirection.

> Hence the PR is to improve the HTTP client redirect logic to allow non-safe HTTP method (POST, PUT, ...) redirection for the 307(Temporary Redirect) and 308(Permanent Redirect) response status codes. 

> Add new config field  `boolean allowAuthHeaders = false;` is introduced and By default Authorization, Proxy-Authenticate, Proxy-Authorization, and WWW-Authenticate headers are removed from the redirect requests. Send them if it set to true during the redirection

Fixes #24681

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: stdlib/http/src/main/ballerina/src/http/redirect/redirect_client.bal [hunk 1]
```diff
@@ -365,36 +378,40 @@ function createNewEndpointConfig(ClientConfiguration config) returns ClientConfi
...
-    } else {
-        return ();
     }
+    if (statusCode == STATUS_TEMPORARY_REDIRECT || statusCode == STATUS_PERMANENT_REDIRECT ||
+               statusCode == STATUS_MULTIPLE_CHOICES || statusCode == STATUS_USE_PROXY) {
+        return httpVerb;
+    }
+    log:printDebug(function() returns string {
+            return "unsupported redirect status code" + statusCode.toString();
+        });
```
Comment (ldclakmal) [inline stdlib/http/src/main/ballerina/src/http/redirect/redirect_client.bal:402]:
```suggestion
        return "unsupported redirect status code" + statusCode.toString();
    });
```

File: tests/jballerina-integration-test/src/test/resources/http/src/httpservices/11_http_redirects.bal [hunk 2]
```diff
@@ -252,6 +253,78 @@ service testRedirect on serviceEndpoint3 {
...
     }
+
+    @http:ResourceConfig {
+        methods: ["GET"],
+        path: "/doPost"
+    }
+    resource function PostClearText(http:Caller caller, http:Request request) {
+        http:Client endPoint4 = new("http://localhost:9103", endPoint4Config );
+        http:Request req = new;
+        req.setHeader("proxy-authorization", "Basic YWxhZGRpbjpvcGVuc2VzYW1l");
```
Comment (ldclakmal) [inline tests/jballerina-integration-test/src/test/resources/http/src/httpservices/11_http_redirects.bal:264]:
```suggestion
        req.setHeader("Proxy-Authorization", "Basic YWxhZGRpbjpvcGVuc2VzYW1l");
```

File: tests/jballerina-integration-test/src/test/resources/http/src/httpservices/11_http_redirects.bal [hunk 3]
```diff
@@ -252,6 +253,78 @@ service testRedirect on serviceEndpoint3 {
...
+        }
+    }
+
+    @http:ResourceConfig {
+        methods: ["GET"],
+        path: "/doSecurePut"
+    }
+    resource function testSecurePut(http:Caller caller, http:Request request) {
+        http:Request req = new;
+        req.setHeader("proxy-authorization", "Basic YWxhZGRpbjpvcGVuc2VzYW1l");
```
Comment (ldclakmal) [inline tests/jballerina-integration-test/src/test/resources/http/src/httpservices/11_http_redirects.bal:286]:
```suggestion
        req.setHeader("Proxy-Authorization", "Basic YWxhZGRpbjpvcGVuc2VzYW1l");
```

File: stdlib/http/src/main/ballerina/src/http/redirect/redirect_client.bal [hunk 4]
```diff
@@ -365,36 +378,40 @@ function createNewEndpointConfig(ClientConfiguration config) returns ClientConfi
...
-    if (statusCode == STATUS_SEE_OTHER) {
-        redirectRequest.removeHeader(TRANSFER_ENCODING);
-        redirectRequest.removeHeader(CONTENT_LENGTH);
+function createRedirectRequest(Request request, boolean allowAuthorizationHeader) returns Request {
+    if (allowAuthorizationHeader) {
+        return request;
     }
-    return redirectRequest;
+    request.removeHeader(AUTHORIZATION);
+    request.removeHeader("Proxy-Authenticate");
```
Comment (ldclakmal) [inline stdlib/http/src/main/ballerina/src/http/redirect/redirect_client.bal:411]: Shall we add these headers as constants?

File: stdlib/http/src/main/ballerina/src/http/redirect/redirect_client.bal [hunk 5]
```diff
@@ -365,36 +378,40 @@ function createNewEndpointConfig(ClientConfiguration config) returns ClientConfi
...
-        string[] headerValues = request.getHeaders(headerName);
-        foreach var headerValue in headerValues {
-            redirectRequest.addHeader(headerName, headerValue);
-        }
-    }
-    if (statusCode == STATUS_SEE_OTHER) {
-        redirectRequest.removeHeader(TRANSFER_ENCODING);
-        redirectRequest.removeHeader(CONTENT_LENGTH);
+function createRedirectRequest(Request request, boolean allowAuthorizationHeader) returns Request {
+    if (allowAuthorizationHeader) {
```
Comment (ldclakmal) [inline stdlib/http/src/main/ballerina/src/http/redirect/redirect_client.bal:407]:
```suggestion
function createRedirectRequest(Request request, boolean allowAuthHeaders) returns Request {
    if (allowAuthHeaders) {
```

File: tests/jballerina-integration-test/src/test/resources/http/src/httpservices/11_http_redirects.bal (hunk 3 above)
Comment (chamil321) [reply tests/jballerina-integration-test/src/test/resources/http/src/httpservices/11_http_redirects.bal:286]: Keeping 1 in lower case to get it tested :)

**Pull request merged**